### PR TITLE
Generate kotlin extension functions for building models

### PR DESCRIPTION
**Pull request by elihart**
https://github.com/airbnb/epoxy/issues/270

I also bumped support to Android Studio 3.0, which required tweaking quite a bit of things to make everything work.

### CODE REVIEW

File: epoxy-processor/src/main/java/com/airbnb/epoxy/KotlinExtensionWriter.kt [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        models.map { buildExtensionForModel(it) }
+                .forEach { fileBuilder.addFun(it) }
+
+        return fileBuilder.build()
+    }
+
+    private fun buildExtensionForModel(model: GeneratedModelInfo): FunSpec {
+        val constructorIsNotPublic = model.constructors
+                .filter { it.params.isEmpty() }
+                .any { !it.modifiers.contains(Modifier.PUBLIC) }
```
Comment (geralt-encore) [inline epoxy-processor/src/main/java/com/airbnb/epoxy/KotlinExtensionWriter.kt]: could be changed to more idiomatic Kotlin:
`Modifier.PUBLIC !in modifiers`

File: epoxy-processor/src/main/java/com/airbnb/epoxy/KotlinExtensionWriter.kt [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+fun String.lowerCaseFirstLetter(): String {
+    if (isEmpty()) {
+        return this
+    }
+
+    return Character.toLowerCase(get(0)) + substring(1)
+}
+
+typealias JavaClassName = com.squareup.javapoet.ClassName
```
Comment (geralt-encore) [inline epoxy-processor/src/main/java/com/airbnb/epoxy/KotlinExtensionWriter.kt]: wow, it is a perfect use-case for typealiases  👍

File: epoxy-processor/src/main/java/com/airbnb/epoxy/KotlinExtensionWriter.kt (hunk 1 above)
Comment (elihart) [reply epoxy-processor/src/main/java/com/airbnb/epoxy/KotlinExtensionWriter.kt]: great point

File: epoxy-processor/src/main/java/com/airbnb/epoxy/KotlinExtensionWriter.kt (hunk 2 above)
Comment (elihart) [reply epoxy-processor/src/main/java/com/airbnb/epoxy/KotlinExtensionWriter.kt]: yeah :) I did it without them at first with FQN's and it was terrible 

Thanks for looking!

**Pull request merged**